Storage Quota imposition on Realtime tables

shounakmk219 opened the pull request:
# Description

This PR allows pinot to impose storage quota restrictions on realtime tables.
To block the ingestion on realtime tables below consideration are kept in mind

1. Replicas should be consistent.
2. Ingestion should auto resume upon quota increase or freed up storage.
3. Easy observability into the blocked ingestion with relevant info. 
4. Query results should be consistent

Also to keep it simple the ingestion is blocked only when storage quotas are breached

# Blocking ingestion

- The storage quota will be imposed as part of the `RealtimeSegmentValidationManager` controller job.
- If the table is breaching quota, the `PauseState` is updated accordingly by setting pause `reasonCode` as `STORAGE_QUOTA_EXCEEDED` on the table IS
- Pause state from table IS is checked during the new consuming segment creation
- If table is paused then new consuming segments will not be created
- Ongoing segment commit will be unaffected.

## What if the `pauseState` is updated manually on zk directly to resume the table ingestion?

Updating the `pauseState` alone will not create the new consuming segments. New consuming segments can be created by:

1. Calling the `/resumeConsumption` API on controller
2. Running `RealtimeSegmentValidationManager` job with setting `recreateDeletedConsumingSegment` as `true`

The resume consumption API internally depends on the `RealtimeSegmentValidationManager` task itself to handle the consuming segment creation. 
As the storage quota validation is itself part of the `RealtimeSegmentValidationManager` job it will again check for the storage quota and set the `pauseState` to the right value.

# Resuming ingestion

Once a table exceeds the storage quota, we may need to resume the consumption in below cases:

1. Storage quota is extended for the table
2. Storage frees up upon older segments being deleted (by retention manager)

To automate the resume consumption flow we will depend on the `RealtimeSegmentValidationManager` job itself. 
In case the user does not want to wait for the periodic job run, they can manually trigger the job from the controller API or use the resume consumption endpoint.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1696,6 +1705,20 @@ private IdealState updatePauseStatusInIdealState(String tableNameWithType, boole
+        ZNRecord znRecord = idealState.getRecord();
+        znRecord.setSimpleField(IS_QUOTA_EXCEEDED, Boolean.valueOf(quotaExceeded).toString());
+        return new IdealState(znRecord);
+      }, RetryPolicies.noDelayRetryPolicy(1));
```
Why not allow retry here?

shounakmk219 replied:
No strong reason to not retry. Was keeping it inline with the table pause IS update. 
Right now in case of failure the next segment commit or validation job, whichever runs first, will update the IS.
Let me know if you have any magic number for retry in mind, maybe 3?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -108,8 +110,20 @@ protected void processTable(String tableNameWithType, Context context) {
+
     _llcRealtimeSegmentManager.ensureAllPartitionsConsuming(tableConfig, streamConfig,
-        context._recreateDeletedConsumingSegment, context._offsetCriteria);
+        context._recreateDeletedConsumingSegment || idealStateUpdated, context._offsetCriteria);
```
We should recreate deleted consuming segment only with case 2 above?

shounakmk219 replied:
That's right. This will effectively recreate the consuming segment only in case 2. 
Even though `idealStateUpdated` is set in case 1 as well, eventually it will skip the consuming segment creation as in case 1 the `IS_QUOTA_EXCEEDED` flag is set which will effectively skip creating consuming segment.
But then I need to ensure the `updateStorageQuotaExceededInIdealState` should return false incase the IS update fails as in that case `IS_QUOTA_EXCEEDED` flag is not set which will end up recreating the consuming segment. Will add that check as well. 
Thanks for pointing it out!

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/StorageQuotaChecker.java:
```diff
@@ -75,9 +80,12 @@ public static QuotaCheckerResponse failure(String msg) {
+    if (!_controllerConf.getEnableStorageQuotaCheck()) {
+      return success("Storage quota check is disabled, skipping the check");
+    }
+    Preconditions.checkArgument(_timeout > 0, "Timeout value must be > 0, input: %s", _timeout);
```
Move this check to the constructor?

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/StorageQuotaChecker.java:
```diff
@@ -205,4 +213,48 @@ public QuotaCheckerResponse isSegmentStorageWithinQuota(TableConfig tableConfig,
     }
   }
+
+  public boolean isTableStorageQuotaExceeded(TableConfig tableConfig) {
```
Should we also check if quota check is enabled?
Consider extracting the common part for `isSegmentStorageWithinQuota()` and this method

shounakmk219 replied:
Agree, will do.

Jackie-Jiang reviewed (commented):
Mostly good. Can we add a test to verify this?

@swaminathanmanish @sajjad-moradi please also take a look

sajjad-moradi commented:
> Mostly good. Can we add a test to verify this?
> 
> @swaminathanmanish @sajjad-moradi please also take a look

Will review next week.

swaminathanmanish commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1696,6 +1705,32 @@ private IdealState updatePauseStatusInIdealState(String tableNameWithType, boole
+   * @return true if the IS was successfully updated for the table. Returns false in case of no op or the update fails.
+   */
+  public boolean updateStorageQuotaExceededInIdealState(String tableNameWithType, boolean quotaExceeded) {
+    IdealState is = getIdealState(tableNameWithType);
```
Discussed with @shounakmk219 to explore stateless approach, where we always read table size, instead of persisting in Zk. 

Ideal state update read-modify-write is expensive we will incur large number of writes when there are large number of segments. This state is also updated from RealtimeSegmentValidation thread on the controller. 

cc @Jackie-Jiang

shounakmk219 replied:
Notes from discussion with @swaminathanmanish 
Need to maintain the state for below reasons:

The ideal state update happens only during the point where table storage exceeds the quota, making it a less frequent operation (not affected by segment commit rate)
- All further segment commits will just undergo read operation (IS is already being fetched during segment commit so not additional read is made for quota exceed check).
- All prior segment commits will only perform size calculation and no IS operation (reuse already available IS)

Whereas in stateless approach we will keep querying all servers to get the table size for all the further segment commits which feels unnecessary.


The resume consumption flow gets tricky with stateless as we don’t have a pointer on whether the consuming segments were not created due to quota breach or were manually deleted.
- Right now the segment validation task does not recreate consuming segments. Unless we have a context on whether the quota was earlier breached or not, we would end up recreating the missing consuming segment at every run which is not the existing behaviour.
- Also this job runs every hour so the IS update operation done by this job won’t be a concern.

swaminathanmanish replied:
> Notes from discussion with @swaminathanmanish Need to maintain the state for below reasons:
> 
> The ideal state update happens only during the point where table storage exceeds the quota, making it a less frequent operation (not affected by segment commit rate)
> 
> * All further segment commits will just undergo read operation (IS is already being fetched during segment commit so not additional read is made for quota exceed check).
> * All prior segment commits will only perform size calculation and no IS operation (reuse already available IS)
> 
> Whereas in stateless approach we will keep querying all servers to get the table size for all the further segment commits which feels unnecessary.
> 
> The resume consumption flow gets tricky with stateless as we don’t have a pointer on whether the consuming segments were not created due to quota breach or were manually deleted.
> 
> * Right now the segment validation task does not recreate consuming segments. Unless we have a context on whether the quota was earlier breached or not, we would end up recreating the missing consuming segment at every run which is not the existing behaviour.
> * Also this job runs every hour so the IS update operation done by this job won’t be a concern.

Thanks @shounakmk219. This makes sense.

shounakmk219 commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1696,6 +1705,32 @@ private IdealState updatePauseStatusInIdealState(String tableNameWithType, boole
+        return new IdealState(znRecord);
+      }, RetryPolicies.noDelayRetryPolicy(1));
+      if (is == null) {
+        LOGGER.error("Failed to set 'isQuotaExceeded' to {} in the Ideal State for table {}.", quotaExceeded,
```
That's right.

swaminathanmanish reviewed (approved):
LGTM

shounakmk219 commented:
@sajjad-moradi can you please review the PR?
Thanks!

sajjad-moradi commented:
I just had a discussion with @mcvsubbu about this. Here's what we think:
With the proposed solution in this PR, there will be two IS flags indicating if the table is paused: "isTablePaused" and "isQuotaExceeded". That doesn't sound right. Instead we can have a more detailed field called pauseStatus. Here's an example for the storage quota exceed case:
```json
"pauseStatus": {
  "isPaused": true,
  "reasonCode": "STORAGE_QUOTA_EXCEEDED",
  "comment": "Current storage 2.3G exceeds quota 2.0G",
  "time": "2024-08-01T09:30:00Z"
}
```
And here is an example for a regular pause that's caused by an admin hitting the pause endpoint:
```json
"pauseStatus": {
  "isPaused": true,
  "reasonCode": "ADMINISTRATIVE",
  "comment": "Need to change the underlying topic",
  "time": "2024-08-01T09:30:00Z"
}
```
This way in case in future there's another reason for pausing the table, there's no need to add yet another field to IS to pause the table. We can simply do that by adding another `reasonCode` which is an enum in the source code.
Moreover, in case in future we want to extend pause functionality like adding support for partition level pausing, we can simply add more fields to pauseStatus.

shounakmk219 commented:
@sajjad-moradi @mcvsubbu , the `pauseStatus` field makes sense. Do you think we need to maintain the set of different pause reasons active on a table as well? Context will be lost during the reason being overridden/not recorded based on how we define the priority of these reasons. 

- paused consumption should not be overridden by quota breach as in that case consumption will resume once quota gets honoured
- quota breach can be overridden by paused consumption as anyways quota limit is checked every time before resuming consumption.

Here the priority is well established so may not be a big concern.

sajjad-moradi commented:
The priority you described makes sense. Basically storage quota should not be even be checked if the table is paused by an admin.

sajjad-moradi commented:
On a separate note, @mcvsubbu thinks it's better to have all the logic for hanlding storage quota (pausing/resuming) in one place, in "realtime segment validation manager" periodic job. 
Of course the benefit is that all the logic is in one place, so it's better in terms of maintainability. The downside of doing that is the the storage won't be checked till the next round of the periodic job. Maybe that's ok because normally consuming segments take hours to complete, and even checking quota upon segment completion is also delayed. WDYT?

sajjad-moradi commented:
@Jackie-Jiang any thoughts here?

mcvsubbu commented:
> On a separate note, @mcvsubbu thinks it's better to have all the logic for hanlding storage quota (pausing/resuming) in one place, in "realtime segment validation manager" periodic job. Of course the benefit is that all the logic is in one place, so it's better in terms of maintainability. The downside of doing that is the the storage won't be checked till the next round of the periodic job. Maybe that's ok because normally consuming segments take hours to complete, and even checking quota upon segment completion is also delayed. WDYT?

There is enough going on during the segment completion, that I would hate to add more logic there with more conditions. Can you state why you need the quota to be checked during segment completion?

I am even fine adding logic in the background task to remove latest segments (and adjust consumption offsets, of course) if the quota is exceeded (it may matter in some cases that the excess quota will result in charges). On the other hand, if this is the case, the metric on current storage use should be monitored in order to flag such situations early and in advance

shounakmk219 commented:
Hey @mcvsubbu @sajjad-moradi , it is ok if the table storage spills over the quota by few segments before stopping the consumption and that said we can keep the quota validation + IS update logic at one place in segment validation manager. 
Also as per the suggestion of introducing `pauseStatus`, a single check  on `pauseStatus` will cover both the pause reasons, eliminating the need of adding any extra checks during segment completion.

> Can you state why you need the quota to be checked during segment completion?

There is no strict storage quota imposition requirement, but also didn’t saw any major concern in putting the check during segment completion

> remove latest segments (and adjust consumption offsets, of course) if the quota is exceeded

We don’t want to do this as this will affect the query results (moving back in time).

If it turns out that segment validation manager job frequency is not working for the storage quota imposition we can introduce a separate job when needed.
Let me know if this sounds good.

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/metrics/ControllerGauge.java:
```diff
@@ -160,7 +160,8 @@ public enum ControllerGauge implements AbstractMetrics.Gauge {
 
-  TABLE_REBALANCE_IN_PROGRESS("tableRebalanceInProgress", false);
+  TABLE_REBALANCE_IN_PROGRESS("tableRebalanceInProgress", false),
+  TABLE_STORAGE_QUOTA_EXCEEDED("tableStorageQuotaExceeded", true);
```
We should not need a metric saying quota is exceeded. Instead, let us have a metric that emits the percent of quota used by the table. Alerts can be set in specific destinations on the value of this metric. Some installations can choose to set the alert at 80%, others at 90%, etc. depending on the action they want to take.

The value can also be greater than 100%, btw.

shounakmk219 replied:
I have removed the `tableStorageQuotaExceeded` gauge. Will let the downstream system figure out the logic based on the `TABLE_QUOTA` and `TABLE_TOTAL_SIZE_ON_SERVER` gauges.

shounakmk219 commented:
Hey @mcvsubbu @sajjad-moradi I have updated the PR with the changes from PauseState PR and also removed the storage quota check from segment commit process and now only have it as part of the `RealtimeSegmentValidationManager`.
Please take another look!
Thanks

sajjad-moradi reviewed (approved):
Thanks for making all the changes.

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -108,8 +112,32 @@ protected void processTable(String tableNameWithType, Context context) {
+
     _llcRealtimeSegmentManager.ensureAllPartitionsConsuming(tableConfig, streamConfig,
-        context._recreateDeletedConsumingSegment, context._offsetCriteria);
+        context._recreateDeletedConsumingSegment || !pauseState.isPaused(), context._offsetCriteria);
```
Please change this to  _not_ call `ensureAllPartitionsConsuming` if the table is paused (or storage is exceeeded) rather than overloading the paused boolean with recreation of deleted segment.

shounakmk219 commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/StorageQuotaChecker.java:
```diff
@@ -43,14 +44,19 @@ public class StorageQuotaChecker {
     _controllerMetrics = controllerMetrics;
     _leadControllerManager = leadControllerManager;
     _pinotHelixResourceManager = pinotHelixResourceManager;
+    _controllerConf = controllerConf;
```
Makes sense, will make the change.

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -108,8 +112,34 @@ protected void processTable(String tableNameWithType, Context context) {
+    // Skips updating the pause state if table is paused by admin
+    PauseState pauseState = computePauseState(tableNameWithType);
+
+    if (!pauseState.isPaused()) {
```
Sorry for the nit, but this reads like it is invoking the periodic consumption fix if the table is not paused. That is partially true, but another factor is that the table may have run out of quota.

So, it is better to create a method like `shouldEnsurePartitionsConsuming()` from which we return a boolean instead of returning the PauseState.

Thanks.

shounakmk219 replied:
Had a chat offline, will handle the future of `_recreateDeletedConsumingSegment` flag in a separate PR.

mcvsubbu replied:
I learned that the flag is not quite as I had imagined. Yes, it is a config, and if turned true, then it ends up invoking the periodic job with 'true' (which ends up handling a condition of someone accidentally or intentionally removing the consuming segments of a table).

the flag is _also_ used in the "resume" part because resume happens via kicking off the periodic task.

cc: @sajjad-moradi

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -108,8 +112,36 @@ protected void processTable(String tableNameWithType, Context context) {
+      boolean unPausedUponStorageWithinQuota =
+          pauseState.getReasonCode().equals(PauseState.ReasonCode.STORAGE_QUOTA_EXCEEDED);
+      _llcRealtimeSegmentManager.ensureAllPartitionsConsuming(tableConfig, streamConfig,
+          context._recreateDeletedConsumingSegment || unPausedUponStorageWithinQuota, context._offsetCriteria);
```
Let us re-factor this as follows:
- Create a method called `shouldRestartConsumption()` that returns a boolean
- Change the original call to `ensureConsuming` to be under an if check based on the return of this method
- The method `shouldRestartConsumption()` returns false if either one of the conditions hold: (a) the table has been paused by admin (2) the table is out of quota

Agree?

shounakmk219 replied:
what about `_recreateDeletedConsumingSegment `?

mcvsubbu commented on pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:
```diff
@@ -108,8 +112,45 @@ protected void processTable(String tableNameWithType, Context context) {
+        pauseState.getReasonCode().equals(PauseState.ReasonCode.STORAGE_QUOTA_EXCEEDED);
+      if (unPausedUponStorageWithinQuota) {
+        // recreate consuming segments if table is resumed upon the table storage getting within quota limit
+        context._recreateDeletedConsumingSegment = true;
```
```suggestion
        return true;
```

shounakmk219 replied:
Once the quota is increased on table or storage is freed due to retention, how will the consuming segments be recreated unless we set this flag?

mcvsubbu replied:
Fair enough. I was thinking of the world when the flag is not there. Maybe it is good to remove the flag in this PR. Alternatively, you can commit this and either one of us can remove the flag.

here is what I have for the flag removal:
- change ensureConsuming() method to be in favor of the flag always being true
- Remove the flag from the config and the map. Anyone who deletes the consuming segments by mistake will want to restart consumption, so it will work. Anyone who deliberately wants to discard consuming segments can do with the pause consumption operation command we have. If necessary, we can enhance the pause command to optionally discard current consuming segments.

I will give a ship it, let me know if you want me to raise a PR to remove that flag.

shounakmk219 replied:
https://github.com/apache/pinot/pull/14024
